server_tgi.py

```diff
@@ -0,0 +1,284 @@
+import argparse
+import json
+import logging
+import time
+import uuid
+from contextlib import asynccontextmanager
+from pathlib import Path
+from typing import AsyncGenerator, Optional, Tuple
+
+import docker
+import requests
+import uvicorn
+from fastapi import FastAPI, Request
+from fastapi.responses import StreamingResponse
+from huggingface_hub import InferenceClient
+from transformers import AutoTokenizer
+
+from functionary.inference_stream import generate_openai_format_from_stream_async
+from functionary.openai_types import (
+    ChatCompletionChunk,
+    ChatCompletionRequest,
+    ChatCompletionResponse,
+    ChatCompletionResponseChoice,
+    ChatMessage,
+    Function,
+    StreamChoice,
+    Tool,
+    UsageInfo,
+)
+from functionary.prompt_template import get_prompt_template_from_tokenizer
+from functionary.prompt_template.prompt_utils import enforce_tool_choice
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def get_health():
+    try:
+        response = requests.get(f"{args.tgi_endpoint}/health")
+        if response.status_code == 200:
+            return "healthy"
+    except requests.RequestException:
+        pass
+
+    return "unhealthy"
+
+
+@asynccontextmanager
+async def lifespan(app: FastAPI):
+    # Run TGI docker container at start up
+    logger.info("Starting TGI docker container...")
+    port = int(args.tgi_endpoint[args.tgi_endpoint.rindex(":") + 1 :])
+    tgi_docker_client = docker.from_env()
+    tgi_container = tgi_docker_client.containers.run(
+        "ghcr.io/huggingface/text-generation-inference:2.0.4",
+        f"--model-id {args.model} --max-total-tokens {args.tgi_max_total_tokens}",
```

Comment:
Instead of starting a docker container always, add a parameter to the CLI so it allows connecting to that server instead of starting a container always

Reply:
Or use that same args.tgi_endpoint, if its set already, connect to it, if its not set, automatically create a new server